### PR TITLE
Feat: Script to Initialize a Cloned User Project

### LINKED ISSUE #1104: Add a script to initialize a cloned user project (like setupRepo)
## This is:
- Feature request

## Description
We need to add a script similar to our [scripts/setupRepo.js](https://github.com/webiny/webiny-js/blob/master/scripts/setupRepo.js) to the user's project, so that new team members can quickly get the project ready for development when cloned from VCS.

Part of this problem is described in our #1099 issue and I think it would be very beneficial to include such a tool out of the box in our `cwp-template-*` packages.

### PR DESCRIPTION
**Pull request by TommyJackson85**
## Related Issue

Closes #1104

## Your solution

Creating a new script to run a project setup script through the packages folder.
the new package is called ```setup-project```, so far it consists of an ```index``` file with a code snippet that was used to run a project setup.
Further work needs to be done to this index file and I am looking for general feedback on my file setup.

## How Has This Been Tested?
Working on this.

## Screenshots (if relevant):
None yet.

### CODE REVIEW

**Comment by TommyJackson85** [comment]
@Pavel910 is there any examples I can look at to see how a .env file was generated in the Webiny project?
Give me some feedback on my current setup anyway!

**Comment by Pavel910** [comment]
@TommyJackson85 here's a reference: `packages/cwp-template-full/template`, and the same thing can be found in `packages/cwp-template-cms/template` (see `index.js` files for actual code that operates on those files). You'll see `example.env.json` files, and they get renamed to `.env.json`. But - that's also one thing we need to fix - the original `example.env.json` should stay in the project, and based on them you can then generate new `.env.json` files when `setup-project` is executed.

File: package.json [hunk 1]
```diff
@@ -138,6 +138,7 @@
     "prepublishOnly": "node scripts/prepublishOnly.js",
     "prettier": "prettier --config .prettierrc.js --write",
     "prettier:check": "prettier \"packages/**/src/**/*.{js,ts,tsx}\" --check --config .prettierrc.js",
+    "setup-project": "node packages/setup-project",
     "setup-repo": "node scripts/setupRepo",
```
Comment (Pavel910) [inline package.json:142]: Currently we use this `setup-repo` script, and we want to keep it that way. The `setup-project` package only needs to be introduced in user's projects, generated by `cwp-template-full` and `cwp-template-cms`.

**Review by Pavel910 — CHANGES_REQUESTED** [review]
@TommyJackson85 so to recap: `setup-project` is a package that needs to be generated in the project created by `create-webiny-project`, and that means that package will be located in the `cwp-template-full` and `cwp-template-cms`.

**Comment by TommyJackson85** [comment]
> @TommyJackson85 here's a reference: `packages/cwp-template-full/template`, and the same thing can be found in `packages/cwp-template-cms/template` (see `index.js` files for actual code that operates on those files). You'll see `example.env.json` files, and they get renamed to `.env.json`. But - that's also one thing we need to fix - the original `example.env.json` should stay in the project, and based on them you can then generate new `.env.json` files when `setup-project` is executed.

Oh I see that the index isn't in the the template folder, Its in the ```template```'s parent folders. The template folders hold the example.env.json files for the index to reference. Ok I will look through the other examples to see how it is implemented.

Will I create a template folder to hold the `example.env.json` file like the examples? Maybe we might add more files to that folder at a later stage.

File: package.json (hunk 1 above)
Comment (TommyJackson85) [reply package.json:142]: Ok so its still good that I included the "setup-project" call in the main package.json file?

Comment (Pavel910) [reply package.json:142]: Our main repo doesn't need that, as there won't be such a package. That package will be generated for the user's project, and will be defined in the template. You need to add that script to the `cwp-template-full/template/dependencies.json` - that file is converted to `package.json` during `create-webiny-project` execution.

**Comment by Pavel910** [comment]
@TommyJackson85 the whole `template` folder gets copied over and tweaked. It is literally the structure of the project you want user to see after `create-webiny-project`.

For `example.env.json` - there is a section in the code that renames the necessary files, but we need to keep these files + create an `.env.json` based on them in the user's project (this is only related to `example.env.json files`).

The example files will then be used by your `setup-project` package to generate the new `.env.json` when another user clones the project.

**Comment by TommyJackson85** [comment]
@Pavel910 
so from ```cwp-template-full``` files (index and the example.env.json) I see this process:
Screenshot 2020-08-03 at 16 55 49
Screenshot 2020-08-03 at 16 55 25

I will copy over the template folder (from ```cwp-template-cms```) into our new folder so.

**Comment by Pavel910** [comment]
@TommyJackson85 During `setup-project` you already have the project, so you don't need to do anything with the template. What I want to achieve is: during `create-webiny-project`, modify the file renaming so that it leaves the `example.env.json` inside the newly generated project. That's it.

Then, in your `setup-project`, simply use those files to generate the `.env.json`.

**Comment by TommyJackson85** [comment]
So are we to change the cwp template index files (i.e. ```packages/cwp-template-full/index``` , ```packages/cwp-template-cms/index```) so that they copy ```example.env.json``` and then alter the copy instead of the original?

**Comment by Pavel910** [comment]
@TommyJackson85 correct :) You can maybe introduce another flag in `filesToCopy.js`, like `keepOriginal: true`, and simply update the for loop that does the copying/renaming to take that flag into consideration.

**Comment by TommyJackson85** [comment]
> @TommyJackson85 correct :) You can maybe introduce another flag in `filesToCopy.js`, like `keepOriginal: true`, and simply update the for loop that does the copying/renaming to take that flag into consideration.

Ah ok, I thought ```filesToCopy.js``` might help haha!

**Comment by TommyJackson85** [comment]
@Pavel910 if keepOriginal was set to True, would we remove the the oldFile ref changes through something like this instead in the loop?
```
            fs.moveSync(
                //removeing oldFile reference
                //path.join(root, filesToCopy[i].dir, filesToCopy[i].oldFile),
                path.join(root, filesToCopy[i].dir, filesToCopy[i].newFile),
                { overwrite: true }
            );
```

**Comment by TommyJackson85** [comment]
@Pavel910 on a side not, do you know what the best is to test this manually?

**Comment by Pavel910** [comment]
@TommyJackson85 I think this is the easiest way; if `keepOriginal === true` - use `copyFileSync`, otherwise use the current `moveSync` to rename the file (I also made the code easier to read so feel free to copy this and modify if necessary):

```js
const filesToCopy = require("./filesToCopy");
for (let i = 0; i < filesToCopy.length; i++) {
  const { dir, oldFile, newFile, keepOriginal } = filesToCopy[i];
  const oldFilePath = path.join(root, dir, oldFile);
  const newFilePath = path.join(root, dir, newFile);

  if (keepOriginal) {
    fs.copyFileSync(oldFilePath, newFilePath);
  } else {
    fs.moveSync(oldFilePath, newFilePath, { overwrite: true });
  }
}
```

**Comment by TommyJackson85** [comment]
@Pavel910 I made your suggested changes to the cwp-template index files, can you check them for me? i will delete those console logs later.

What about ```setup-project/index```. Is there any further changes to be done with this file? Are the cwp-template index files running it already?

From running ```node webiny/webiny-js/packages/create-webiny-project test-project --template=webiny/webiny-js/packages/cwp-template-full``` in an outside folder I seem to be getting a package error which I think is unrelated. Will I create a separate issue for it?
```
open-source-projects git:(master) ✗ node webiny/webiny-js/packages/create-webiny-project test-project --template=webiny/webiny-js/packages/cwp-template-full

  IMPORTANT NOTICE:
  ----------------------------------------
  We've detected a global installation of @webiny/cli. This might not play well with your new project.
  We recommend you do one of the following things:

   - uninstall the global @webiny/cli package by running npm rm -g @webiny/cli or
   - run webiny commands using yarn webiny so that the package is always resolved to your project dependencies

  The second option is also recommended if you have an older version of Webiny project you want to keep using.
  ----------------------------------------

  Creating project at /Users/thomasjackson/Desktop/software-projects/open-source-projects/test-project:
  ✔ Pre-template setup
  ✖ Install template package
    → info Visit https://yarnpkg.com/en/docs/cli/add for documentation about thi
…
    Create project folders
    Set up project dependencies
    Initialize git
    Get proper package versions
    Resolve packages
    Install dependencies
    Run template-specific actions

    ERROR OUTPUT:
    ----------------------------------------
    Command failed with exit code 1: yarn add --exact @webiny/cwp-template-webiny/webiny-js/packages/cwp-template-full@latest --cwd /Users/thomasjackson/Desktop/software-projects/open-source-projects/test-project
    error Couldn't find package "@webiny/cwp-template-webiny/webiny-js/packages/cwp-template-full" on the "npm" registry.
    yarn add v1.22.4
    info No lockfile found.
    [1/4] Resolving packages...
    info Visit https://yarnpkg.com/en/docs/cli/add for documentation about this command.
    ----------------------------------------

    Please open an issue including the error output at https://github.com/webiny/webiny-js/issues/new.
    You can also get in touch with us on our Slack Community: https://www.webiny.com/slack


  Writing log to /Users/thomasjackson/Desktop/software-projects/open-source-projects/cwp-logs.txt...
  Cleaning up project...
  Project cleaned!
```

**Comment by TommyJackson85** [comment]
@Pavel910 I just realised that in each ```filesToCopy.js```, an object already exists that copies ```example.env.json``` to ```.env.json``` but it has a directory of "api" (mine newly added objects dir was left blank as you seen in the file changes):
 ```  
{
        dir: "api",
        oldFile: "example.env.json",
        newFile: ".env.json",
        keepOriginal: false
    },
```
Is this for another purpose? If not will I remove the object I added and just edit this object's ```keepOriginal``` file to true

**Comment by Pavel910** [comment]
Hey @TommyJackson85, sorry, I somehow missed your previous question :( Lets go one by one:

The error `Couldn't find package "@webiny/cwp-template-webiny/webiny-js/packages/cwp-template-full"`:
This is caused by not adding `./` to the path. Make sure to start local paths with `./` so the tools knows you're referencing a local folder, and not a remote package. See how it generated the package name of `@webiny/cwp-template-webiny/webiny-js/packages/cwp-template-full`, that's because you need to pass `--template=./webiny/....`.


Regarding the `filesToCopy`:
We basically want to always keep the `example.env.json` file, no matter where it is located. It means that you'll have one in project root (thus the `dir: ""`), one in `api` folder, one in `apps/admin`, etc.

**Comment by TommyJackson85** [comment]
> @TommyJackson85 it's missing for obvious reason: this PR is supposed to implement that script. Read the related issue for more details.
> 
> Part 1 of the solution is to keep `example.env.json` files - that's now done. Now we need to move to the second part of the issue, and that's implementing the actual ` setup-project` tool itself. 

@Pavel910 I am looking at the related Issue #1104 , I am just still trying to grasp how this is working but I am learning :)

When running ```node ./webiny/webiny-js/packages/create-webiny-project test-project --template=./webiny/webiny-js/packages/cwp-template-full```, it generates a copy of the 'cwp-template-full' ```template``` folder as ```test-project```, in which it also renames some of the files and makes copys of env files, etc. ***am I right?***

If so, Should I add a ```packages``` folder to ```cwp-template-full/template``` folder, with the ```setup-project``` package inside that packages folder so its reachable?

The ```setup-project ``` package (which holds the code snippet you provided in the related issue) is already added to the root directory packages. I think this was meant to be included in the ```template``` folders instead.

**Comment by Pavel910** [comment]
> When running node ./webiny/webiny-js/packages/create-webiny-project test-project --template=./webiny/webiny-js/packages/cwp-template-full, it generates a copy of the 'cwp-template-full' template folder as test-project, in which it also renames some of the files and makes copys of env files, etc. am I right?

Yes, that is correct.

> If so, Should I add a packages folder to cwp-template-full/template folder, with the setup-project package inside that packages folder so its reachable?

That was the idea, yes. We want to setup that package for developers out of the box. It will allow developers to modify it for their specific projects, and it will setup the base for all other project packages (by having the `packages` folder). The `setup-project` should be located in the `cwp-template-full/template` and all the other `cwp-*` packages. Of course the code of the actual `setup-project` needs to be modified according to the project template.

**Comment by TommyJackson85** [comment]
> That was the idea, yes. We want to setup that package for developers out of the box. It will allow developers to modify it for their specific projects, and it will setup the base for all other project packages (by having the `packages` folder). The `setup-project` should be located in the `cwp-template-full/template` and all the other `cwp-*` packages. Of course the code of the actual `setup-project` needs to be modified according to the project template..

Ok this is much clearer. I apologise, from the initial issue, I thought you were referring to the webiny project with project root, haha!
I will focus just on ```cwp-package-full``` for the time being. As I will later test the ```setup-project``` script command after creating another test-project the setup-project package included. :) give me a few days though as I am back to my regular job in the morning.

**Comment by Pavel910** [comment]
@TommyJackson85 no worries, we're not in a hurry with this one. We're glad you're taking part in the project 👍

**Comment by TommyJackson85** [comment]
@Pavel910 the setup-project package is now included, and tested, in both cwp template packages. Running ```yarn setup-project``` appears to work in both but the .env.json files appear to exist anyway:
```
 test-project git:(master) ✗ yarn setup-project
yarn run v1.22.4
$ node ./packages/setup-project
✍️  Writing environment config files...
⚠️  .env.json already exists, skipping.
⚠️  api/.env.json already exists, skipping.
⚠️  apps/admin/.env.json already exists, skipping.
🏗  Building packages...
lerna notice cli v3.22.1
lerna info versioning independent
lerna info Executing command in 3 packages: "npm run build"
api-apollo-gateway: > api-apollo-gateway@0.1.0 build /Users/thomasjackson/Desktop/software-projects/open-source-projects/test-project/api/apolloGateway
api-apollo-gateway: > webiny run build
api-settings-manager: > api-settings-manager@0.1.0 build /Users/thomasjackson/Desktop/software-projects/open-source-projects/test-project/api/settingsManager
api-settings-manager: > webiny run build
api-database-proxy: > api-database-proxy@0.1.0 build /Users/thomasjackson/Desktop/software-projects/open-source-projects/test-project/api/databaseProxy
api-database-proxy: > webiny run build
api-apollo-gateway: 2020-08-27T17:35:21.200Z webiny Start bundling
api-database-proxy: 2020-08-27T17:35:21.201Z webiny Start bundling
api-settings-manager: 2020-08-27T17:35:21.203Z webiny Start bundling
api-settings-manager: ● SettingsManager █████████████████████████ compiling (0%)  
api-settings-manager:  

api-apollo-gateway: ✔ ApolloGateway
api-apollo-gateway:   Compiled successfully in 4.05s
api-apollo-gateway: 2020-08-27T17:35:25.402Z webiny Finished bundling
lerna success run Ran npm script 'build' in 3 packages in 5.8s:
lerna success - api-apollo-gateway
lerna success - api-database-proxy
lerna success - api-settings-manager
✅️ Packages were built successfully!

🏁 Your repo is almost ready!
Update .env.json with your MongoDB connection string and you're ready to develop!

✨  Done in 6.85s.
➜  test-project git:(master) ✗ 

```

I also added an additional helpful message in the console log when running the ```create-webiny-project``` package (check the file changes). Its just a quick guide for running setup-project. It's not finished yet, I just included it to get some feedback from you, to see if you like it and on what to write. If you don't want anything written yet, I will remove it.

Is there any additional ways we can test this? Anything else you want to include or remove?

**Comment by TommyJackson85** [comment]
Wrote up descriptive procedures on the setup in ```docs/DEVELOPMENT_OVERVIEW.md```, but please fact check me!

@Pavel910 if you can give my PR another review, it would be great. Let me know what I can remove, add, change, etc. I will move onto another task in the mean time. I wont do any more to this just yet!

**Comment by Pavel910** [comment]
Closing this PR as we've implemented this in v5.

**Pull request closed without merging**